drivers:platform:stm32:Added DMA Driver

jansunil opened the pull request:
Added STM32 DMA platform driver

## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [x] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,99 @@
+	if (!dma_desc)
+		return -ENOMEM;
+
+	ddesc = (struct stm32_dma_desc*)no_os_calloc(1, sizeof(*ddesc));
+	if (!ddesc)
+		return -ENOMEM;
```
why is this needed? after all ddesc is a pointer, which points to the already allocated DMA handler in generated_main.c (I would assume a similar handling is done for DMA as with UART and other peripherals).

rbolboac commented on drivers/platform/stm32/stm32_dma.h:
```diff
@@ -0,0 +1,72 @@
+ * @brief Structure holding the STM32 DMA parameters.
+ */
+struct stm32_dma_init_param {
+	DMA_Stream_TypeDef* dma_stream;
+	uint32_t priority;
+	bool mem_increment;
+	bool per_increment;
+	uint32_t mem_data_alignment;
+	uint32_t periph_data_alignment;
+	uint32_t mode;
+	uint32_t dma_channel;
+	uint32_t direction;
+	enum no_os_irq_event event;
```
the init param can be directly the HandleTypeDef pointer for DMA

jansunil replied:
I've added a handletypedef to the inti params now

rbolboac reviewed (commented):
This still needs some work.
First of all, I would recommend to take a look at how other stm32 peripherals are handled and handle the DMA in a similar way, UART can be such an example where the UART handle in generated_main is used.
Secondly, please take a look at what was done for maxim platform, the API you currently have simply initializes the peripheral, without taking into consideration any of no_os_dma_init_param fields. Those fields should be mapped to the stm32 hdma instance.
For example the instance can be the id, the mode can be mapped to xfer_type and so on.
I would say it's no easy task to add support for dma...and it will require a lot of API implementations and functionality mapping.
Of course, implementation for the other platform ops will be necessary, and I think maxim_dmac.c drivers can serve as an example for this.
Lastly, it would be nice to have a little context, to see how you want to use the DMA.
For example, in no_os there is support for SPI transfers with dma, see no_os_spi_transfer_dma_sync and no_os_spi_transfer_dma_async which were also implemented in maxim_spi.c driver (can serve as an example if this is the final goal for DMA usage).

jansunil commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,99 @@
+ * @brief stm32 platform specific DMA platform ops structure
+ */
+const struct no_os_dma_platform_ops stm32_dma_ops = {
+	.dma_init = stm32_dma_init,
```
Added a remove function now

jansunil commented:
@rbolboac -  Thanks for the suggestions. There are some updates made to the driver in order to use the other structure made available in the no_os_dma_init_params. The driver flow is also aligned to the maxim implementation.

In order to get a context of how the drivers can be used, an example implementation has been added in the stm32_spi_dma_transfer_sync() implementation in the existing SPI driver. 

Please let me know on your inputs further.

buha commented on drivers/platform/stm32/stm32_spi.h:
```diff
@@ -66,6 +75,14 @@ struct stm32_spi_desc {
 	/** Chip select gpio descriptor */
 	struct no_os_gpio_desc *chip_select;
+	/** DMA Descriptor */
+	struct no_os_dma_desc* dma_desc;
```
that's good progress, i see fewer members in this structure but it can be improved further

this field shouldn't be here, actually the parent descriptor of this `stm32_spi_desc` is a `no_os_dma_desc` and you have direct access to that in the DMA API functions

jansunil replied:
@buha  I am not sure I got the point right here. Do you mean to add the no_os_dma_desc to the no_os_spi_desc? We might need access to this DMA descriptor somewhere in the SPI descriptors so that it can be used in the async functions

jansunil commented:
Hi all - are there any further review comments on this PR?

rbolboac commented:
> Hi all - are there any further review comments on this PR?

First request would be to please fix CI errors.

buha commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -45,6 +45,9 @@
 #include "no_os_alloc.h"
+#include "stm32_dma.h"
+#include "no_os_pwm.h"
+#include "stm32_pwm.h"
```
is this pwm dependency really needed ?

jansunil replied:
So the dependency we have is to control the chip select.
Since this is a DMA operation and cannot toggle chip select individually for each byte/word, we need a PWM module here

CiprianRegus replied:
As far as I know this functionality is not supported even for regular SPI transfers. Is there a use case for this? Why wouldn't you chain multiple byte/word transfers and handle the CS control in the transfer done interrupt?

jansunil replied:
The goal of using a DMA is to read huge chunks of samples without interference of the core. The rate at which the read happens is at around 1MSPS, where in I would end up reading 2 bytes per conversion. Triggering interrupts at that high a rate just to control the state of the CS GPIO alone would be too much overhead on the core.

Hence, I am trying to control the CS via a PWM, where in I read n number of samples and get into the interrupt handler after reception of 400 samples and not after every byte

CiprianRegus commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+	int ret;
+	uint8_t i;
+
+	if (!desc)
```
The current DMA driver for maxim uses shared references for `no_os_dma_desc` descriptors (meaning that if you call the dma_init function with the same `id` field you'll get the same `no_os_dma_desc *` pointer). Is there a reason for not implementing the same here?

jansunil replied:
I am not sure I got this right..- what is the shared references you are referring to here?

CiprianRegus commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+
+	no_os_irq_ctrl_remove(desc->irq_ctrl);
+
+	for (i = 0; i < desc->num_ch; i++)
+		no_os_free(&desc->channels[i]);
```
This is incorrect since you'll end up double freeing some memory. Should be `no_os_free(desc->channels)`

jansunil replied:
I've removed free of channels. I suppose freeing up the desc should take care of this as well?

CiprianRegus replied:
You still need to free `desc->channels`. What I was referring to is to replace

```
for (i = 0; i < desc->num_ch; i++)
		no_os_free(&desc->channels[i]);
```

with

```
no_os_free(desc->channels);
```

CiprianRegus commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+ */
+int stm32_dma_remove(struct no_os_dma_desc* desc)
+{
+	int ret;
```
Unused

CiprianRegus commented on drivers/platform/stm32/stm32_dma.h:
```diff
@@ -0,0 +1,106 @@
+ * @struct stm32_dma_channel
+ * @brief STM32 DMA Channels
+ */
+struct stm32_dma_channel {
```
I'm a bit confused by this... Is this struct specific to a DMA instance or channel? Why do we need fields such as `src`, `dst` or `length` during init?

jansunil replied:
This struct is specific to the channel. 
We do not really need them in the init, but its required in the stm32_dma_xfer_start(). The HAL_DMA_Start apis need to be provided with the src/dst addresses in order to initiate the DMA transfer

CiprianRegus commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,6 +216,40 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
+		ret = no_os_dma_init(&sdesc->dma_desc, sinit->dma_init);
+		if (ret) {
+			return ret;
+		}
```
remove the brackets

CiprianRegus commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,6 +216,40 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
+	if (sinit->dma_init) {
+		ret = no_os_dma_init(&sdesc->dma_desc, sinit->dma_init);
+		if (ret) {
+			return ret;
```
`goto error`

CiprianRegus commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -386,12 +424,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
+
+	rx_ch->id = sdma_rx->hdma;
+	rx_ch->sg_list = NULL;
+	if (sdma_rx->irq_num > 0)
```
The `irq_num` is `uint32_t`. Moreover, I think this adds unnecessary complexity. I don't see why would the user care to assign a specific interrupt number.

jansunil replied:
irq_num is a field required in the no_os_irq_register_callback()/ no_os_irq_set_priority(). These are being called in the no_os_dma_config_xfer() for configuring transfer complete callbacks

CiprianRegus commented on projects/ad74413r/src.mk:
```diff
@@ -37,7 +38,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_alloc.c \
-		$(NO-OS)/util/no_os_mutex.c
+		$(NO-OS)/util/no_os_mutex.c \
+		$(NO-OS)/util/no_os_dma.c \
```
No such file

CiprianRegus commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -386,12 +424,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
+ * @param is_async - Whether or not the function should wait for the completion.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t stm32_spi_dma_transfer_sync(struct no_os_spi_desc* desc,
```
What is the behavior of this transfer function? I see it doesn't wait for the transfer to finish before returning, but `callback` is ignored as well.

jansunil replied:
I am trying to implement a non blocking function, the goal is to return from this API irrespective of whether the transfer/receive is complete or not. 
The callback is being registeredwithin the no_os_dma_config_xfer() and enabled via the no_os_dma_xfer_start()

rbolboac commented on drivers/platform/stm32/stm32_dma.h:
```diff
@@ -0,0 +1,106 @@
+	/* DMA Mode */
+	enum stm32_dma_mode dma_mode;
+	/* IRQ ID */
+	uint32_t irq_num;
```
already present in struct no_os_dma_ch, should use that instead

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+				     DMA_MINC_DISABLE;
+	sdma_ch->hdma->Init.PeriphInc = sdma_ch->per_increment ? DMA_PINC_ENABLE :
+					DMA_PINC_DISABLE;
+	sdma_ch->hdma->Init.MemDataAlignment = sdma_ch->mem_data_alignment;
+	sdma_ch->hdma->Init.PeriphDataAlignment = sdma_ch->per_data_alignment;
+	sdma_ch->hdma->Init.Mode = sdma_ch->dma_mode;
```
don't really understand the purpose of these and also the switch statement below. basically they realize the same thing without the invalid value check

jansunil replied:
Intention is to configure the DMA init parameters via the platform drivers. 
A switch case has been added since there are enum elements in the stm32_dma.h to realize the parameters, so that the user doesn't really have to look into the HAL drivers to configure the metrics. Do you suggest any other alternate approach here?

rbolboac replied:
what I mean is that you should remove the 3 lines above, because the switch blocks are assigning MemDataAlignment, PeriphDataAlignment and Mode fields.

jansunil replied:
Oh yeah, you are right here. I had overlooked the duplicated assignments- I've removed the 3 lines above

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+	if (sdma_chan->irq_num) {
+		ret = HAL_DMA_Start_IT(sdma_chan->hdma, (uint32_t)sdma_chan->src,
+				       (uint32_t )sdma_chan->dst, sdma_chan->length);
+		if (ret != HAL_OK)
+			return -EINVAL;
```
can be taken out of if / else

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+	} else {
+		ret = HAL_DMA_Start(sdma_chan->hdma, (uint32_t )sdma_chan->src,
+				    (uint32_t )sdma_chan->dst, sdma_chan->length);
+		if (ret != HAL_OK)
+			return -EINVAL;
```
can be taken out of if / else

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+
+	if (sdma_chan->irq_num) {
+		ret = HAL_DMA_Abort_IT(sdma_chan->hdma);
+		if (ret != HAL_OK)
+			return -EINVAL;
```
can be taken out of if / else

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -0,0 +1,284 @@
+			return -EINVAL;
+	} else {
+		ret = HAL_DMA_Abort(sdma_chan->hdma);
+		if (ret != HAL_OK)
+			return -EINVAL;
```
can be taken out of if / else

rbolboac commented on drivers/platform/stm32/stm32_dma.h:
```diff
@@ -0,0 +1,106 @@
+/***************************************************************************//**
+ *   @file   stm32_dma.h
+ *   @brief  Platform independent function definitions and data types
+ * 	     for the DMA API.
```
missing author

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -191,6 +194,7 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 	csip.pull = NO_OS_PULL_NONE;
 	csip.extra = &csip_extra;
 	csip.platform_ops = &stm32_gpio_ops;
+
```
unrelated change

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,6 +216,40 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 
+	if (sinit->dma_init) {
+		ret = no_os_dma_init(&sdesc->dma_desc, sinit->dma_init);
+		if (ret) {
```
brackets not needed

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,6 +216,40 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
+			return ret;
+		}
+
+		if (sinit->rxdma_ch) {
```
brackets not needed for single line statements, see everywhere where applicable

rbolboac reviewed (commented):
some other comments from my side,
also please squash the CI fix commit into the first one.
I would also suggest to have 2 commits:
- one to add stm32_dma support
- the second one to add dma support for SPI transfers

jansunil commented:
Changelog: 
Addressed comments from @CiprianRegus

jansunil commented:
Hi All- Please hold the review. I shall address all the comments and add a comment here

jansunil commented on drivers/platform/stm32/stm32_dma.h:
```diff
@@ -0,0 +1,106 @@
+	/* DMA Handle */
+	DMA_HandleTypeDef *hdma;
+	/* Channel Number */
+	uint32_t ch_num;
```
no_os_dma_ch consists of the ``id``, which is interpreted as the dma handle in the no_os_dma_config_xfer() - https://github.com/analogdevicesinc/no-OS/blob/main/drivers/api/no_os_dma.c#L285

The ch_num parameter needs to be a 32 bit number.

Please let me know if my understanding is wrong

rbolboac commented on drivers/platform/stm32/stm32_pwm.c:
```diff
@@ -56,6 +56,8 @@
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
+
+#ifdef HAL_TIM_MODULE_ENABLED
```
please remove these from stm32_pwm, it is not necessary

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -229,6 +228,7 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 		sdesc->dma_desc->sg_handler = sinit->dma_init->sg_handler;
 	}
 
+#ifdef HAL_TIM_MODULE_ENABLED
```
all changes added in stm32_spi should be moved in the previous commit

rbolboac commented on projects/max14906/src/platform/stm32/platform_src.mk:
```diff
@@ -10,6 +12,8 @@ INCS += $(PLATFORM_DRIVERS)/stm32_delay.h     \
 SRCS += $(PLATFORM_DRIVERS)/stm32_delay.c     \
 	$(PLATFORM_DRIVERS)/stm32_gpio.c      \
 	$(PLATFORM_DRIVERS)/stm32_spi.c       \
+	$(PLATFORM_DRIVERS)/stm32_pwm.c       \
```
I think the whole idea for using HAL_TIM_MODULE_ENABLED is to not have to  include stm32_pwm.c, unless the ioc contains a timer module and it actually uses PWM for CS. this is not valid for any of the projects is No-OS. so please remove all stm32_pwm files, and no_os_pwm files from projects (added in this commit).
the same remarks applies for dma. currently no project is using spi with DMA, so why would we include those files? stm32_spi should compile without stm32_dma, if HAL_DMA_MODULE_ENABLED is not defined.

of course, you can add a project which showcases how to use SPI with DMA transfers and PWM CS. this will be use as an example (should contain DMA init and PWM init structures and the ioc should contain the modules as well).

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -45,6 +45,10 @@
+#ifdef HAL_DMA_MODULE_ENABLED
+#include "stm32_dma.h"
+#endif
+#include "no_os_pwm.h"
```
include only if HAL_TIM_MODULE_ENABLED is defined

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -60,6 +60,7 @@ static struct no_os_dma_desc *dma_descriptor;
 int stm32_dma_config_xfer(struct no_os_dma_ch *channel,
 			  struct no_os_dma_xfer_desc *xfer)
 {
+#ifdef HAL_DMA_MODULE_ENABLED
```
why would you use stm32_dma.c if HAL_DMA_MODULE_ENABLED is not defined? I think this check is not needed...
only use this check in stm32_spi, because the spi driver can be used without the DMA module.
also, this change is not related to this commit and should be part of the first commit.

rbolboac commented on drivers/platform/stm32/stm32_dma.c:
```diff
@@ -68,7 +69,11 @@ int stm32_dma_config_xfer(struct no_os_dma_ch *channel,
 
 	sdma_ch = channel->extra;
 
+#if defined (STM32L4) || defined (STM32H5)
```
a comment here would be nice... why only these families? I see there are other families which have hdma->Instance available. I think the problem is that the Init.Channels is not available for most of them.
so maybe change this check to only use Init.Channel for stm32f2, f4 and f7: https://github.com/search?q=org%3ASTMicroelectronics+hdma-%3EInit.Channel&type=code
also, this change is not related to this commit and should be part of the first commit.

rbolboac commented on drivers/platform/stm32/stm32_pwm.h:
```diff
@@ -64,6 +64,8 @@ enum TimOCMode {
 	TIM_OC_PWM2 = 2,
 };
 
+#ifdef HAL_TIM_MODULE_ENABLED
```
please remove these from stm32_pwm, it is not necessary. same remark applies as for DMA, why would you use stm32_pwm without a timer module?
there is no HAL_SPI_MODULE_ENABLED check in stm32_spi...

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,9 +216,45 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 	if (ret)
 		goto error;
 
+	if (sinit->dma_init) {
```
everything related to DMA from stm32_spi shoul use #ifdef HAL_DMA_MODULE_ENABLED and everything related to PWM should use HAL_TIM_MODULE_ENABLED

rbolboac reviewed (commented):
it still needs a little bit of work from my point of view.

CiprianRegus commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -386,12 +430,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
+int32_t stm32_spi_dma_transfer_async(struct no_os_spi_desc* desc,
+				     struct no_os_spi_msg* msgs,
+				     uint32_t len,
+				     void (*callback)(void*),
+				     void* ctx, bool is_async)
```
Same as before, this callback is totally ignored and the `stm32_spi_dma_transfer_async` function doesn't take the same parameters as `no_os_spi_transfer_dma_async`, so there is no way this can work.

jansunil replied:
The callback is being registeredwithin the no_os_dma_config_xfer() and enabled via the no_os_dma_xfer_start()

jansunil replied:
@CiprianRegus - what is the difference between the callback that's in the argument here and the one in the no_os_dma_init param?

CiprianRegus replied:
The callback in the `no_os_dma_init_param` struct is meant to be called once a DMA transfer finished (doesn't matter if the transfer was for the SPI peripheral or it was a memory to memory transfer). The callback in the `no_os_spi_transfer_dma_async()` function should be called once the DMA SPI transfer issued by the function call is done.

jansunil replied:
As discussed in the call, I've added a condition to overwrite the desc->sg_handler if the user provides a callback in the function argument of the ``no_os_spi_transfer_dma_async()`` .

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -212,9 +218,47 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
+		/* Initialize CS PWM */
+		ret = no_os_pwm_init(&sdesc->pwm_desc, sinit->pwm_init);
+		if (ret)
+			goto error;
```
missing no_os_dma_remove in case of an error

rbolboac commented on drivers/platform/stm32/stm32_spi.c:
```diff
@@ -386,12 +430,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
 const struct no_os_spi_platform_ops stm32_spi_ops = {
 	.init = &stm32_spi_init,
 	.write_and_read = &stm32_spi_write_and_read,
 	.remove = &stm32_spi_remove,
```
the spi remove should remove everyting that was added in spi_config (dma and pwm)

rbolboac reviewed (commented):
also, second commit message has a typo, replace 4) with 2)

jansunil commented:
Hi All- The builds are all passing now. I have also addressed to the comments and replied back to the conversations. Does this look good to go?

rbolboac commented on projects/adxrs290-pmdz/src.mk:
```diff
@@ -29,7 +29,8 @@ INCS += $(INCLUDE)/no_os_delay.h        \
         $(INCLUDE)/no_os_alloc.h        \
-        $(INCLUDE)/no_os_mutex.h
+        $(INCLUDE)/no_os_mutex.h         \
+        $(INCLUDE)/no_os_pwm.h
```
is this needed?

jansunil replied:
The adxrs290-pmdz project has the TIM module enabled in the .ioc file. Hence added the PWM drivers here as we are making a check for the HAL_TIM_MODULE_ENABLED in the stm32_spi.c driver

rbolboac commented on projects/adxrs290-pmdz/src.mk:
```diff
@@ -41,4 +42,5 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c     \
         $(NO-OS)/util/no_os_list.c      \
         $(NO-OS)/util/no_os_util.c      \
         $(NO-OS)/util/no_os_alloc.c     \
+        $(DRIVERS)/api/no_os_pwm.c     \
```
is this needed?

jansunil replied:
The adxrs290-pmdz project has the TIM module enabled in the .ioc file. Hence added the PWM drivers here as we are making a check for the HAL_TIM_MODULE_ENABLED in the stm32_spi.c driver

rbolboac commented on projects/adxrs290-pmdz/src/platform/stm32/platform_src.mk:
```diff
@@ -2,13 +2,17 @@ INCS += $(PLATFORM_DRIVERS)/stm32_delay.h     \
         $(PLATFORM_DRIVERS)/stm32_gpio.h      \
         $(PLATFORM_DRIVERS)/stm32_hal.h       \
         $(PLATFORM_DRIVERS)/stm32_spi.h       \
+        $(PLATFORM_DRIVERS)/stm32_pwm.h       \
```
is this needed?

jansunil replied:
The adxrs290-pmdz project has the TIM module enabled in the .ioc file. Hence added the PWM drivers here as we are making a check for the HAL_TIM_MODULE_ENABLED in the stm32_spi.c driver

rbolboac commented on projects/eval-adis1646x/src.mk:
```diff
@@ -26,7 +26,7 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_units.h \
 		$(INCLUDE)/no_os_alloc.h \
-        	$(INCLUDE)/no_os_mutex.h
+        $(INCLUDE)/no_os_mutex.h
```
unrelated change

jansunil replied:
Removed this

rbolboac commented on projects/eval-adis1647x/src.mk:
```diff
@@ -26,7 +26,7 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_util.h \
 		$(INCLUDE)/no_os_units.h \
 		$(INCLUDE)/no_os_alloc.h \
-        	$(INCLUDE)/no_os_mutex.h
```
unrelated change

jansunil replied:
Removed this

rbolboac commented on projects/eval-adxl313z/src.mk:
```diff
@@ -26,7 +26,7 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_init.h \
 		$(INCLUDE)/no_os_alloc.h \
-        	$(INCLUDE)/no_os_mutex.h
+        $(INCLUDE)/no_os_mutex.h
```
unrelated change

jansunil replied:
Removed this

rbolboac commented on projects/eval-adxl313z/src.mk:
```diff
@@ -38,7 +38,7 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
 		$(NO-OS)/util/no_os_alloc.c \
-        	$(NO-OS)/util/no_os_mutex.c
+        $(NO-OS)/util/no_os_mutex.c
```
unrelated change

rbolboac commented on projects/eval-adxl355-pmdz/src.mk:
```diff
@@ -27,7 +27,7 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_units.h \
 		$(INCLUDE)/no_os_init.h \
 		$(INCLUDE)/no_os_alloc.h \
-        	$(INCLUDE)/no_os_mutex.h
```
unrelated change

jansunil replied:
Removed this

rbolboac commented on projects/eval-adxl355-pmdz/src.mk:
```diff
@@ -40,7 +40,7 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
 		$(NO-OS)/util/no_os_alloc.c \
-        	$(NO-OS)/util/no_os_mutex.c
+        $(NO-OS)/util/no_os_mutex.c
```
unrelated change

jansunil replied:
Removed this

rbolboac reviewed (commented):
also, the third commit should be squashed in the second one because the changes depend on each other.

jansunil commented:
Changelog:
1) Removed unrelated spacing changes in the src/stm32_src.mk
2) Added snippet to overwrite the callback function if the user specifies one via the no_os_spi_transfer_dma_async()

jansunil commented:
Changelog:
Fixed LinuxBuilds Documentation errors

rbolboac commented on projects/eval-adxl355-pmdz/src.mk:
```diff
@@ -40,7 +40,7 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
 		$(NO-OS)/util/no_os_alloc.c \
-        	$(NO-OS)/util/no_os_mutex.c
+		$(NO-OS)/util/no_os_mutex.c
```
for some reason these whitespace changes still appear here in the commit

rbolboac reviewed (approved):
looks good from my point of view, although @CiprianRegus should have the final word.
also, please do not forget to add (in the near future) a new PR with a project example using STM32 DMA with SPI (with and without PWM for CS would be nice) and the necessary documentation on wiki.

jansunil commented:
Changelog: Addressed offline comment from @CiprianRegus - Removed the is_async parameter from the ``stm32_spi_dma_transfer_async()`` in order to match the no_os_spi.h definition
